ZOOKEEPER-2940. Deal with maxbuffer as it relates to large requests from clients

anmolnar opened the pull request:
https://issues.apache.org/jira/browse/ZOOKEEPER-2940

This PR covers the other part of Jute buffer monitoring which relates to client response sizes.

`jute.maxbuffer` is often set too small or the default (4MB) value is not enough to receive large responses from the server which causes the client unable to operate. e.g. `getChildren()` on a node which has lots of children or execution of a large multi.

These new metrics lets operators to monitor the size of generated responses to get an idea on how to properly set client's `jute.maxbuffer`.

The other part of monitoring which relates to proposal size has already been merged https://github.com/apache/zookeeper/pull/415

rakeshadr commented on src/java/main/org/apache/zookeeper/server/command/MonitorCommand.java:
```diff
@@ -75,9 +79,9 @@ public void commandRun() {
-            print("last_proposal_size", leader.getProposalStats().getLastProposalSize());
-            print("max_proposal_size", leader.getProposalStats().getMaxProposalSize());
-            print("min_proposal_size", leader.getProposalStats().getMinProposalSize());
+            print("last_proposal_size", leader.getProposalStats().getLast());
```
Infact, getLastProposalSize, getMaxProposalSize, getMinProposalSize is more meaningful. Can we keep this names?

anmolnar replied:
Sure, no problem. I thought it was redundant to refer the class name in method names, but I'll revert that.

rakeshadr replied:
One more observation. It seems "ProposalStats" is reused for clientResponseStats metrics, but it may create confusions due to the method name mismatches. Again, ProposalStats javadocs says "Provides live statistics about a running Leader."  Also, stats will evolve and could be chance of adding unrelated metrics later. 

How about create a new class "ResponseStats" and name metrics like below. If we look at the existing min_proposal_size metrics, they didn't use the term leader/quorum. Keeping that in mind, do we need specifically ''client'' term in the metrics, simply response gives a context to the users that server-to-client response. Whats your opinion?
```
last_response_size,
max_response_size,
min_response_size
```

anmolnar replied:
Agreed, javadoc is misleading.

Actually I wanted to keep these two things together and would like to refactor both in a later commit. They refer to the same thing basically, exposing statistics on Jute buffer usage which is a feed of int values. I'm happy to find a better name for the class, but wouldn't create a separate one for almost the same purpose.

What do you think of `BufferStats`?

anmolnar replied:
You're right about stats will evolve with time, but proposal and client request size are common in Jute buffer usage. I have a strong feeling that these two things will always evolve together hence it makes sense to keep the stats together which also makes refactoring easier. Later, if it turns out that they diverge, we can easily split them.

rakeshadr commented on src/java/test/org/apache/zookeeper/server/NIOServerCnxnTest.java:
```diff
@@ -68,5 +74,19 @@ public void testOperationsAfterCnxnClose() throws IOException,
+            zk.create("/a", "test".getBytes(), Ids.OPEN_ACL_UNSAFE,
+                    CreateMode.PERSISTENT);
+
+            assertThat(stats.getLast(), greaterThan(0));
```
0, looks like magic number. Any possibility to give a non-zero value, matching closer to the expected value ?

Also, please add error message - assertThat("error message", stats.getLast(), greaterThan(0));

anmolnar replied:
+1 for assert messages
Did you mean -1 as the magic number which means that stat has not been updated yet...?

rakeshadr replied:
please add comments about the expectations like, 
assertEquals("stat should be initialized with -1", -1, stats.getLast());  or a better message:)

that will help easy test case maintenance.
b) greaterThan(0) , similarly good to add a message explaining why 0?

rakeshadr commented on src/java/test/org/apache/zookeeper/server/NettyServerCnxnTest.java:
```diff
@@ -84,4 +92,17 @@ public void testSendCloseSession() throws Exception {
+            zk.create("/a", "test".getBytes(), Ids.OPEN_ACL_UNSAFE,
+                    CreateMode.PERSISTENT);
+
+            assertThat(stats.getLast(), greaterThan(0));
```
Same comment as above...non-zero expected value.

rakeshadr commented on src/java/main/org/apache/zookeeper/server/admin/Commands.java:
```diff
@@ -329,12 +329,20 @@ public CommandResponse run(ZooKeeperServer zkServer, Map<String, String> kwargs)
             response.put("open_file_descriptor_count", osMbean.getOpenFileDescriptorCount());
             response.put("max_file_descriptor_count", osMbean.getMaxFileDescriptorCount());
 
+            response.put("last_client_response_size", stats.getClientResponseStats().getLast());
```
Please update zookeeperAdmin.xml about these commands. Please refer monitoring the health of the cluster section in this page.

rakeshadr replied:
Also, I could see contrib/monitoring/ganglia/zookeeper_ganglia.py file contains metrics section. Probably you could refer 'zk_open_file_descriptor_count': {'units': 'descriptors'}, as an example case.

anmolnar commented:
@rakeshadr I updated the pull request with your code review comments:
- stats class has been renamed,
- added comments to unit test asserts,
- replaced initial value magic number with a constant.
Please take a look.

rakeshadr commented on src/java/main/org/apache/zookeeper/server/quorum/Leader.java:
```diff
@@ -106,10 +104,10 @@ public String toString() {
         new HashSet<LearnerHandler>();
 
-    private final ProposalStats proposalStats;
+    private final BufferStats bufferStats;
```
Could you use "proposalStats" variable name instead of "bufferStats", thats more readable.

rakeshadr commented on src/java/main/org/apache/zookeeper/server/command/StatCommand.java:
```diff
@@ -64,8 +64,8 @@ public void commandRun() {
                 Leader leader = ((LeaderZooKeeperServer)zkServer).getLeader();
-                ProposalStats proposalStats = leader.getProposalStats();
-                pw.printf("Proposal sizes last/min/max: %s%n", proposalStats.toString());
+                BufferStats bufferStats = leader.getBufferStats();
```
Same as above. Please use var name "proposalStats" instead of "bufferStats"

anmolnar commented:
@rakeshadr thanks, this is done.

phunt commented:
is this intended for 3.4 or just 3.5+ ? The jira says 3.5+, is that right?

anmolnar commented on src/contrib/monitoring/ganglia/zookeeper_ganglia.py:
```diff
@@ -213,7 +213,13 @@ def metric_init(params=None):
         'zk_synced_followers': {'units': 'nodes'},
-        'zk_pending_syncs': {'units': 'syncs'}
+        'zk_pending_syncs': {'units': 'syncs'},
+        'zk_last_proposal_size': {'units': 'ms'},
```
Right. Good catch.

anmolnar commented on src/docs/src/documentation/content/xdocs/zookeeperAdmin.xml:
```diff
@@ -1897,6 +1896,12 @@ server.3=zoo3:2888:3888</programlisting>
               zk_pending_syncs    0               - only exposed by the Leader
               zk_open_file_descriptor_count 23    - only available on Unix platforms
               zk_max_file_descriptor_count 1024   - only available on Unix platforms
+              zk_last_proposal_size
```
Correct, I'll add some example values.

anmolnar commented:
@phunt This is definitely for 3.4 too. I updated the Jira and will create a separate pull request for 3.4.

phunt commented:
@anmolnar should we apply the same logic as my recent comment on #466 ? Add this to jmx and the mntr command output, but don't add it to the legacy/text 4lw. What do you think.

anmolnar commented:
@phunt Sure, just let me try to clarify what needs to be done, because I think you wanted to link #415 instead of #466 which is this PR.
- stats should be added to JMX (done, already in ZooKeeperBean),
- stats should be added to stat and mntr 4lw commands (also done)
What am I missing? What's legacy exactly?
Also would we like to add Jetty support here, I've seen it mentioned in the other PR?

eribeiro commented:
@anmolnar Patrick is mentioning the plan to retire 4lw CLI commands in favor of Restful web service version (based on Jetty). See https://issues.apache.org/jira/browse/ZOOKEEPER-1346

anmolnar commented:
@eribeiro Thanks. In which case given that this PR is targeted for the master branch, I'll add them to Jetty and JMX commands.

eribeiro commented:
Cool. :)
FYI, https://issues.apache.org/jira/browse/ZOOKEEPER-2693
Particularly, this quote:
_"A simple patch would provide a configuration option that disables these 4lw by default. We should include this in 3.4/3/5 code lines. Additionally in 3.5 Jetty metrics support was added, replacing the legacy 4lw. In 3.5.3 and later we should disable 4lw entirely (through a second config option) and encourage people to use Jetty instead."_

phunt commented:
Currently listed as having a conflict, @anmolnar can you update the patch?

anmolnar commented:
@phunt Rebased and conflicts resolved.

phunt commented:
+1, lgtm. Also tried running against a runtime cluster and it worked ok. committed to 3.5.5 and master. Thanks @anmolnar .
